pkg/registry/core/pod/storage/eviction.go

```diff
@@ -130,13 +130,55 @@ func (r *EvictionREST) Create(ctx context.Context, name string, obj runtime.Obje
 
 	// Evicting a terminal pod should result in direct deletion of pod as it already caused disruption by the time we are evicting.
 	// There is no need to check for pdb.
-	if pod.Status.Phase == api.PodSucceeded || pod.Status.Phase == api.PodFailed {
-		_, _, err = r.store.Delete(ctx, eviction.Name, rest.ValidateAllObjectFunc, deletionOptions)
-		if err != nil {
-			return nil, err
+	if canIgnorePDB(pod) {
+		deleteRefresh := false
+		continueToPDBs := false
+		// Preserve current deletionOptions if we need to fall through to checking PDBs later.
+		preservedDeletionOptions := deletionOptions.DeepCopy()
+		err := retry.RetryOnConflict(EvictionsRetry, func() error {
```

Comment:
it's not completely obvious to me why we need retries on conflicts.  Is this conflicting often enough that it warrants the complexity?  I worry that the intent of this PR was simply to add an ` || pod.Status.phase == pending` and it's gotten stuck in review over complicated retry logic we didn't need.

Reply:
Previously, the decision about whether a pod could be evicted was based on a terminal state, so no subsequent update to the pod could invalidate that decision.

This PR adds the possibility of basing that decision on something that can be changed by an update, and introduces the possibility of a race between eviction deleting the pod and the node starting the pod. Example flow:
1. pod is pending
2. eviction request is submitted
2. eviction handler decides to evict based on the pod pending
3. pod is scheduled
4. node observes the scheduled pod and starts the process
5. eviction handler deletes the pod

That means we need to add a precondition to ensure the eviction is performed on the same data we based our decision on. That introduces the possibility of conflict errors, which we should handle internally since clients submitting unconditional evictions aren't expecting to receive conflicts.

Reply:
> That introduces the possibility of conflict errors, which we should handle internally since clients submitting unconditional evictions aren't expecting to receive conflicts.

This bit needs to be a comment in the code.  Also, it seems that conflicts from deletes are reasonable given finalizers, but if conflict retry is important to you, so be it.